doc/iuser.8
.TH IUSER 8 "Version 1.0: 23 Feb 2017"
.SH NAME
ipmiutil_user \- handle user functions

.SH SYNOPSIS
.B "ipmiutil user [-cmsx -NUPREFTVY] <function>

.SH DESCRIPTION
This
.I ipmiutil user
subcommand handles IPMI LAN user functions separately.
These functions can also be handled from ipmiutil lan -q, -u, -p, -v.

This utility can use either the /dev/ipmi0 driver from OpenIPMI,
the /dev/imb driver from Intel, the /dev/ipmikcs driver from valinux,
direct user-space IOs, or the IPMI LAN interface if \-N.

.SH OPTIONS
Command line options are described below.
.IP "-c"
Canonical, delimited output format
.IP "-m 002000"
Target a specific MC (e.g. bus 00, sa 20, lun 00).
This could be used for PICMG or ATCA blade systems.
The trailing character, if present, indicates SMI addressing if 's',
or IPMB addressing if 'i' or not present.
.IP "-s"
When getting info, also get the DCMI sensor information.
.IP "-x"
Causes extra debug messages to be displayed.
.IP "-N nodename"
Nodename or IP address of the remote target system.  If a nodename is
specified, IPMI LAN interface is used.  Otherwise the local system
management interface is used.
.IP "-U rmt_user"
Remote username for the nodename given.  The default is a null username.
.IP "-P/-R rmt_pswd"
Remote password for the nodename given.  The default is a null password.
.IP "-E"
Use the remote password from Environment variable IPMI_PASSWORD.
.IP "-F drv_t"
Force the driver type to one of the followng:
imb, va, open, gnu, landesk, lan, lan2, lan2i, kcs, smb.
Note that lan2i means lan2 with intelplus.
The default is to detect any available driver type and use it.
.IP "-J"
Use the specified LanPlus cipher suite (0 thru 17): 0=none/none/none,
1=sha1/none/none, 2=sha1/sha1/none, 3=sha1/sha1/cbc128, 4=sha1/sha1/xrc4_128,
5=sha1/sha1/xrc4_40, 6=md5/none/none, ... 14=md5/md5/xrc4_40.
Default is 3.
.IP "-T"
Use a specified IPMI LAN Authentication Type: 0=None, 1=MD2, 2=MD5, 4=Straight Password, 5=OEM.
.IP "-V"
Use a specified IPMI LAN privilege level. 1=Callback level, 2=User level, 3=Operator level, 4=Administrator level (default), 5=OEM level.
.IP "-Y"
Yes, do prompt the user for the IPMI LAN remote password.
Alternatives for the password are \-E or \-P.

.SH FUNCTIONS

.IP "list [channel]"
List the IPMI LAN users with enabled status, privilege level, and username.
Optionally specify a channel number.

.IP "enable <user_num> [channel]"
Enable the specified user number and optionally specify an IPMI LAN channel number.
.IP "disable <user_num>]"
Disable the specified user number, so that it has 'No access'.

.IP "set <user_num> name <username>"
Set the username for a specified user number.

.IP "set <user_num> password <password>"
Set the password for a specified user number.

.IP "set <user_num> priv <priv_level> [channel]"
Set the privilege level for a specified user number, where the privilege level is 4=Admin, 3=Operator, 2=User.  Also optionally specify a channel number.

.IP "help"
Show the help (usage) message

.SH "EXAMPLES"
ipmiutil user set 2 name user2 
.br
Sets the username for user 2 to "user2". 


.SH "SEE ALSO"
ipmiutil(8) ialarms(8) iconfig(8) idiscover(8) ievents(8) ifru(8) igetevent(8) ihealth(8) ilan(8) ireset(8) isel(8) isensor(8) iserial(8) isol(8) iwdt(8)

.SH WARNINGS
See http://ipmiutil.sourceforge.net/ for the latest version of ipmiutil and any bug fix list.

.SH COPYRIGHT
Copyright (C) 2017  Andy Cress
.PP
See the file COPYING in the distribution for more details
regarding redistribution.
.PP
This utility is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY.

.SH AUTHOR
.PP
Andy Cress <arcress at users.sourceforge.net>
.br



NEWS
IPMIUTIL NEWS 

In October 2001, this project started as part of the Carrier Grade Linux effort.
It was then known as 'panicsel' ("http://panicsel.sf.net") and included a kernel patch to write a Linux panic event to the IPMI firmware log, as well as utilities.  

In July 2002, this code was first included in MontaVista CGE Linux 2.1.
The panicsel functionality was included in OSDL CGL 1.0 and 2.0 requirements.
The kernel panic functionality was included in the OpenIPMI driver for 2.6 Linux kernels, and in the Intel IMB Linux driver v28 and greater.

A Linux rpm and a Windows setup package for ipmiutil is included on 
the platform CD with Intel carrier-grade servers.

The ipmiutil (or panicsel) rpm is known to be included in the following 
Linux distributions:
MontaVista CGE 2.1, 3.0, 3.1, 4.0, SuSE SLES9, SLES10, RedFlag 5.0

See http://ipmiutil.sourceforge.net for binaries, documentation, etc.

In Oct 2001 (0.8.1), started panicsel.
In Dec 2001 (1.0.0), released with Intel Telecom Linux Technology project
In Jul 2002 (1.2.1), released with MontaVista CGE 2.1
In Jan 2004 (1.4.1), added support for Windows 2000 & 2003
In Aug 2004 (1.4.8), the project was moved from panicsel.sf.net to ipmiutil.sf.net.  The new name more clearly reflects the purpose of the project in its current state.  
In Nov 2004 (1.5.0), support for IPMI LAN was added using FreeIPMI.
In May 2006 (1.7.0), built-in IPMI LAN support was added.
In Aug 2006 (1.7.5), driverless direct I/O support was added
In Jan 2007 (1.8.0), added ipmiutil meta-command, IPMI LAN 2.0 (lanplus) support
In Feb 2007 (1.9.2), added Linux Serial-Over-Lan console feature 
In Oct 2007 (2.0.0), added Windows SOL console and remote soft-shutdown features
In Mar 2008 (2.1.0), the Windows SOL client application was completed.
In Aug 2008 (2.2.0), iconfig function was added to save/restore the BMC configuration parameters
In Sep 2008 (2.2.1), support for the Microsoft IPMI driver was added.
In Nov 2008 (2.3.0), support for the Sun Solaris bmc driver was added.
In Mar 2009 (2.3.5), support for the FreeBSD OS was added.
In Oct 2009 (2.5.0), added ANSI VT100  emulation for Windows SOL console
In Feb 2010 (2.6.0), new subcommand naming scheme (i+subcmd)
In Apr 2010 (2.6.2), ifruset was added to set all FRU Product Area fields
In Jun 2010 (2.6.5), added picmg and firewall functionality
In Aug 2010 (2.6.8), Fedora 14 version of ipmiutil submitted
In Sep 2010 (2.6.9), added fwum/hpm functionality, and OEM Fujitsu sensor/sel 
In Sep 2010 (2.7.0), added sunoen, ekanalyzer, fru_picmg, oem_kontron
In Sep 2010 (2.7.1), added ppc64, mips64, sparc arch support
In Nov 2010 (2.7.2), added Debian package support
In Nov 2010 Fedora 14 released with ipmiutil-2.6.8
In Dec 2010 (2.7.3), added SuperMicro OEM functions
In Mar 2011 OpenSuSE 11.4 released with ipmiutil-2.7.2
In May 2011 (2.7.6), added VLAN support and WIN64 fixes
In Sep 2011 (2.7.8), added IPv6, systemd, Dell OEM functions, PICMG 2.3, tsol
In Oct 2011 (2.7.9), support Wind River and ARM cross-compile, Intel S2600
In Dec 2011 (2.8.0), added DCMI 1.1 commands, devel pkg, MSI package
In Feb 2012 (2.8.1), added Quanta S99Q functions, Dell fixes, Intel ME fixes
In Mar 2012 (2.8.2), added Intel Romley OEM support
In Jun 2012 (2.8.4), added support for HP-UX builds and DCMI 1.1 commands
In Oct 2012 (2.8.6), added pre-built Windows DLL for sample development apps
In Feb 2013 (2.8.8), added much more SuperMicro OEM support
In Apr 2013 (2.9.0), added iseltime to sync sys/RTC/SEL times
In May 2013 (2.9.1), added support for MacOSX client
In Dec 2016 (3.0.1), added support for Lenovo OEM sensors
In Feb 2017 (3.0.1), added support for ASUS OEM sensors
In Mar 2017 (3.0.3), added iuser to consolidate user functions



util/iuser.c
/*
 * iuser.c
 * Handle IPMI user command functions
 *
 * Change history:
 *  03/22/2017 ARCress - included in source tree
 *
 *----------------------------------------------------------------------*
The BSD License 

Copyright (c) 2017, Andy Cress. All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are met:

  a.. Redistributions of source code must retain the above copyright notice, 
      this list of conditions and the following disclaimer. 
  b.. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation 
      and/or other materials provided with the distribution. 
  c.. Neither the name of Intel Corporation nor the names of its contributors 
      may be used to endorse or promote products derived from this software 
      without specific prior written permission. 

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND 
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR 
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON 
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *----------------------------------------------------------------------*/
#ifdef WIN32
#include <windows.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "getopt.h"
#else
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#if defined(HPUX)
/* getopt is defined in stdio.h */
#elif defined(MACOS)
/* getopt is defined in unistd.h */
#include <unistd.h>
#else
#include <getopt.h>
#endif
#endif
#include "ipmicmd.h"

/* global variables */
#ifdef METACOMMAND
extern char * progver;  /*from ipmiutil.c*/
static char * progname  = "ipmiutil user";
#else
static char * progver   = "3.08";
static char * progname  = "iuser";
#endif
static char   fdebug    = 0;
static char   fcanonical = 0;
static char   bdelim = '|';
static uchar  g_bus  = PUBLIC_BUS;
static uchar  g_sa   = BMC_SA;
static uchar  g_lun  = BMC_LUN;
static uchar  g_addrtype = ADDR_SMI;
extern int verbose;  /*see ipmilanplus.c*/

extern void lprintf(int level, const char * format, ...);  /*ipmilanplus.c*/

/* from ilan.c */
char *parse_priv(uchar c);
int DisableUser (int unum, uchar chan);
int SetPasswd(int unum, char *uname, char *upswd, uchar chan, uchar priv);
int GetBmcEthDevice(uchar lan_parm, uchar *pchan);
int GetUserInfo(uchar unum, uchar chan, uchar *enab, uchar *priv, char *uname, char fdebug);

static void printf_user_usage(void)
{
	printf("ipmiutil user Commands:\n");
	printf("\tuser list [channel]\n");
	printf("\tuser enable <user_num> [channel]\n");
	printf("\tuser disable <user_num>\n");
	printf("\tuser set <user_num> name <username> \n");
	printf("\tuser set <user_num> password <password> \n");
	printf("\tuser set <user_num> priv <priv_level> [channel]\n");
	printf("\t\twhere priv_level: 4=Admin, 3=Operator, 2=User\n");
}

/* ipmi_user_list - print out info for users
 *
 * @intf:	ipmi inteface
 * @argc:	argument count
 * @argv:	argument list
 * returns 0 on success, -1 on error
 */
static int ipmi_user_list(void * intf, int  argc, char ** argv)
{
	int i,ret = 0;
	int rv = -1;
	uchar num = 1;
	uchar chan = 1;
	uchar enab, priv;
	char uname[21];
	char *privstr;

	if (argv[0] != NULL) chan = atoi(argv[0]);
	else i = GetBmcEthDevice(0xFF, &chan);
	while (ret == 0) {
		uname[0] = '\0';
		ret = GetUserInfo(num, chan, &enab, &priv, uname, fdebug);
		if (ret == 0) {
			privstr = parse_priv(priv);
			if (fcanonical) 
				printf("User %2d %c chan=%d %c %8s %c %8s %c %s\n",
						num,bdelim,chan,bdelim,(enab? "enabled": "disabled"),
						bdelim,privstr,bdelim,uname);
			else
				printf("User %2d: chan=%d \t%8s \t%8s \t%s\n",
						num,chan,(enab? "enabled": "disabled"),privstr,uname);
			rv = 0;
		}
		num++;
	}
	return(rv);
}

/* ipmi_user_enable_disable  -  enable/disable BMC functions
 *
 * @intf:	ipmi inteface
 * @enable:     whether to enable or disable
 * @argc:	argument count
 * @argv:	argument list
 *
 * returns 0 on success, -1 on error
 */
static int
ipmi_user_enable_disable(void * intf, int enable, int  argc, char ** argv)
{
	int i, ret = -1;
	uchar unum, chan, enab;
	uchar priv = 3;
	char uname[21] = {'\0'};

	if (argc < 1 || strncmp(argv[0], "help", 4) == 0) {
		printf_user_usage();
		return ERR_USAGE;
	}
	if (argv[0] == NULL) { printf_user_usage(); return ERR_USAGE; }
	unum = atoi(argv[0]);

	if (argv[1] != NULL) chan = atoi(argv[1]);
	else i = GetBmcEthDevice(0xFF, &chan);

	if (enable == 0) {
	  ret = DisableUser(unum, chan);
	} else {
	  ret = GetUserInfo(unum, chan, &enab, &priv, uname, fdebug);
	  ret = SetPasswd(unum, uname, NULL, chan, priv);
	}
	return ret;
}

/* ipmi_user_reset - reset firmware user to enable everything
 *
 * @intf:	ipmi inteface
 * @argc:	argument count
 * @argv:	argument list
 *
 * returns 0 on success, -1 on error
 */
static int
ipmi_user_set(void * intf, int  argc, char ** argv)
{
	int ret = -1;
	int i, func = -1;
	char *uname = NULL;
	char *upass = NULL;
	uchar unum, chan, enab, dummy;
	uchar priv = 3;
	char name[21] = {'\0'};

	if ((argc < 2) || (argc > 0 && strncmp(argv[0], "help", 4) == 0)) {
		printf_user_usage();
		return ERR_USAGE;
	}
	if ((argv[1] == NULL) || (argv[2] == NULL)) { 
		printf_user_usage(); 
		return ERR_USAGE; 
	}
	unum = atoi(argv[0]);
	if (strncmp(argv[1], "name", 4) == 0) {
	    func = 1; 
		uname = argv[2];
	} else if (strncmp(argv[1], "password", 8) == 0) {
		func = 2; 
		upass = argv[2];
	} else if (strncmp(argv[1], "priv", 4) == 0) {
	    func = 3; 
		priv = atoi(argv[2]);
	} else {
		printf_user_usage();
		return(ERR_USAGE);
	}
	if (argv[3] != NULL) chan = atoi(argv[3]);
	else i = GetBmcEthDevice(0xFF, &chan);

	switch(func) {
    case 1:  /*set name*/ 
	  	ret = GetUserInfo(unum, chan, &enab, &priv, name, fdebug);
		ret = SetPasswd(unum, uname, upass, chan, priv);
		break;
    case 2:  /*set password*/ 
	  	ret = GetUserInfo(unum, chan, &enab, &priv, name, fdebug);
		if (ret == 0) uname = name;
		ret = SetPasswd(unum, uname, upass, chan, priv);
		break;
    case 3:  /*set priv*/ 
	  	ret = GetUserInfo(unum, chan, &enab, &dummy, name, fdebug);
		if (ret == 0) uname = name;
		ret = SetPasswd(unum, uname, upass, chan, priv);
		break;
	}
	return ret;
}

#ifdef METACOMMAND
int i_user(int argc, char **argv)
#else
#ifdef WIN32
int __cdecl
#else
int
#endif
main(int argc, char **argv)
#endif
{
	void *intf = NULL;
	int rc = 0;
	int c, i;
	char *s1;

	printf("%s ver %s\n", progname,progver);

	parse_lan_options('V',"4",0);  /*default to admin priv*/

    while ( (c = getopt( argc, argv,"cm:T:V:J:EYF:P:N:R:U:Z:x?")) != EOF )
	switch (c) {
          case 'c': fcanonical = 1; break;
          case 'm': /* specific IPMB MC, 3-byte address, e.g. "409600" */
                    g_bus = htoi(&optarg[0]);  /*bus/channel*/
                    g_sa  = htoi(&optarg[2]);  /*device slave address*/
                    g_lun = htoi(&optarg[4]);  /*LUN*/
                    g_addrtype = ADDR_IPMB;
                    if (optarg[6] == 's') {
                             g_addrtype = ADDR_SMI;  s1 = "SMI";
                    } else { g_addrtype = ADDR_IPMB; s1 = "IPMB"; }
		    ipmi_set_mc(g_bus,g_sa,g_lun,g_addrtype);
                    printf("Use MC at %s bus=%x sa=%x lun=%x\n",
                            s1,g_bus,g_sa,g_lun);
                    break;
          case 'x': fdebug = 1; verbose = 1;
			break;  /* debug messages */
          case 'N':    /* nodename */
          case 'U':    /* remote username */
          case 'P':    /* remote password */
          case 'R':    /* remote password */
          case 'E':    /* get password from IPMI_PASSWORD environment var */
          case 'F':    /* force driver type */
          case 'T':    /* auth type */
          case 'J':    /* cipher suite */
          case 'V':    /* priv level */
          case 'Y':    /* prompt for remote password */
          case 'Z':    /* set local MC address */
                parse_lan_options(c,optarg,fdebug);
                break;
          case '?': 
				print_lan_opt_usage(0);
				printf_user_usage();
				return ERR_USAGE;
                break;
	}
    for (i = 0; i < optind; i++) { argv++; argc--; }

	if (argc < 1 || strncmp(argv[0], "help", 4) == 0) {
		printf_user_usage();
	} else if (strncmp(argv[0], "list", 4) == 0) {
		rc = ipmi_user_list(intf, argc-1, &(argv[1]));
	} else if (strncmp(argv[0], "enable", 6) == 0) {
		rc = ipmi_user_enable_disable(intf, 1, argc-1, &(argv[1]));
	} else if (strncmp(argv[0], "disable", 7) == 0) {
		rc = ipmi_user_enable_disable(intf, 0, argc-1, &(argv[1]));
	} else if (strncmp(argv[0], "set", 3) == 0) {
		rc = ipmi_user_set(intf, argc-1, &(argv[1]));
	} else {
		printf_user_usage();
		rc = ERR_BAD_PARAM;
	}
    ipmi_close_();
	return rc;
}
